# Data Understanding 

In [1]:
import pandas as pd 

accounts = pd.read_csv("../data/raw/ravenstack_accounts.csv")
subscriptions = pd.read_csv("../data/raw/ravenstack_subscriptions.csv")
feature_usage = pd.read_csv("../data/raw/ravenstack_feature_usage.csv")
support_tickets = pd.read_csv("../data/raw/ravenstack_support_tickets.csv")
churn_events = pd.read_csv("../data/raw/ravenstack_churn_events.csv")

### Shapes

In [6]:
print(accounts.shape)
print(subscriptions.shape)
print(feature_usage.shape)
print(support_tickets.shape)
print(churn_events.shape)

(500, 10)
(5000, 14)
(25000, 8)
(2000, 9)
(600, 9)


### Samples

In [8]:
accounts.sample(5)

,account_id,account_name,industry,country,signup_date,referral_source,plan_tier,seats,is_trial,churn_flag
313,A-883b7d,Company_313,Cybersecurity,IN,2024-03-28,other,Basic,17,False,False
217,A-e1462e,Company_217,FinTech,UK,2023-03-12,partner,Basic,15,False,False
99,A-503d5a,Company_99,FinTech,US,2023-10-21,other,Enterprise,4,False,True
66,A-526d93,Company_66,HealthTech,FR,2024-04-14,organic,Basic,43,False,False
297,A-a865bd,Company_297,Cybersecurity,US,2023-09-19,organic,Pro,38,False,False


In [9]:
subscriptions.sample(5)

,subscription_id,account_id,start_date,end_date,plan_tier,seats,mrr_amount,arr_amount,is_trial,upgrade_flag,downgrade_flag,churn_flag,billing_frequency,auto_renew_flag
818,S-0509b8,A-e351e1,2024-12-26,2024-12-29,Basic,33,627,7524,False,False,False,True,annual,True
135,S-183073,A-e7a1e2,2024-07-19,NaN,Enterprise,14,2786,33432,False,False,False,False,monthly,True
1865,S-209456,A-95b24a,2023-07-18,NaN,Pro,38,1862,22344,False,False,False,False,annual,True
408,S-a554da,A-712f1c,2024-04-09,2024-07-23,Enterprise,17,3383,40596,False,False,False,True,annual,True
2996,S-7ab68f,A-bcbe98,2024-09-18,NaN,Basic,46,0,0,True,False,False,False,annual,True


In [10]:
feature_usage.sample(5)

,usage_id,subscription_id,usage_date,feature_name,usage_count,usage_duration_secs,error_count,is_beta_feature
22400,U-90b07a,S-66bae1,2024-10-18,feature_25,8,3384,0,False
23337,U-1fbf02,S-937207,2023-09-08,feature_27,11,3938,0,False
14843,U-3d06cd,S-baba71,2024-04-23,feature_14,8,1264,0,True
2488,U-44bed8,S-f16bb2,2024-08-06,feature_23,10,2440,0,True
12240,U-367868,S-2e965d,2023-08-07,feature_10,9,2340,0,True


In [12]:
support_tickets.sample(5)

,ticket_id,account_id,submitted_at,closed_at,resolution_time_hours,priority,first_response_time_minutes,satisfaction_score,escalation_flag
35,T-f68dde,A-118d55,2024-09-08,2024-09-08 14:00:00,14.0,medium,1,4.0,False
985,T-4e07bf,A-600734,2023-10-29,2023-10-29 14:00:00,14.0,urgent,123,3.0,False
767,T-0088b7,A-ce550d,2024-05-15,2024-05-17 17:00:00,65.0,medium,173,3.0,True
1503,T-9fff22,A-b07346,2024-03-08,2024-03-10 04:00:00,52.0,medium,162,NaN,False
1779,T-4ad95c,A-118d55,2024-12-03,2024-12-05 02:00:00,50.0,medium,122,NaN,False


In [13]:
churn_events.sample(5)

,churn_event_id,account_id,churn_date,reason_code,refund_amount_usd,preceding_upgrade_flag,preceding_downgrade_flag,is_reactivation,feedback_text
39,C-2e2a3f,A-7df7a7,2024-07-10,budget,0.00,False,False,False,missing features
117,C-576ecc,A-993085,2024-01-21,features,0.00,False,False,False,NaN
32,C-935f48,A-34ea2f,2024-10-05,unknown,0.00,True,False,False,NaN
556,C-f67026,A-b7ef83,2024-07-24,budget,60.25,True,False,False,missing features
294,C-33fc6a,A-c16101,2024-12-19,features,0.00,False,True,False,NaN


### Data types and non-null count

In [16]:
accounts.info()

<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   account_id       500 non-null    str  
 1   account_name     500 non-null    str  
 2   industry         500 non-null    str  
 3   country          500 non-null    str  
 4   signup_date      500 non-null    str  
 5   referral_source  500 non-null    str  
 6   plan_tier        500 non-null    str  
 7   seats            500 non-null    int64
 8   is_trial         500 non-null    bool 
 9   churn_flag       500 non-null    bool 
dtypes: bool(2), int64(1), str(7)
memory usage: 32.4 KB


In [17]:
subscriptions.info()

<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   subscription_id    5000 non-null   str  
 1   account_id         5000 non-null   str  
 2   start_date         5000 non-null   str  
 3   end_date           486 non-null    str  
 4   plan_tier          5000 non-null   str  
 5   seats              5000 non-null   int64
 6   mrr_amount         5000 non-null   int64
 7   arr_amount         5000 non-null   int64
 8   is_trial           5000 non-null   bool 
 9   upgrade_flag       5000 non-null   bool 
 10  downgrade_flag     5000 non-null   bool 
 11  churn_flag         5000 non-null   bool 
 12  billing_frequency  5000 non-null   str  
 13  auto_renew_flag    5000 non-null   bool 
dtypes: bool(5), int64(3), str(6)
memory usage: 376.1 KB


In [18]:
feature_usage.info()

<class 'pandas.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   usage_id             25000 non-null  str  
 1   subscription_id      25000 non-null  str  
 2   usage_date           25000 non-null  str  
 3   feature_name         25000 non-null  str  
 4   usage_count          25000 non-null  int64
 5   usage_duration_secs  25000 non-null  int64
 6   error_count          25000 non-null  int64
 7   is_beta_feature      25000 non-null  bool 
dtypes: bool(1), int64(3), str(4)
memory usage: 1.4 MB


In [19]:
support_tickets.info()

<class 'pandas.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ticket_id                    2000 non-null   str    
 1   account_id                   2000 non-null   str    
 2   submitted_at                 2000 non-null   str    
 3   closed_at                    2000 non-null   str    
 4   resolution_time_hours        2000 non-null   float64
 5   priority                     2000 non-null   str    
 6   first_response_time_minutes  2000 non-null   int64  
 7   satisfaction_score           1175 non-null   float64
 8   escalation_flag              2000 non-null   bool   
dtypes: bool(1), float64(2), int64(1), str(5)
memory usage: 127.1 KB


In [20]:
churn_events.info()

<class 'pandas.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   churn_event_id            600 non-null    str    
 1   account_id                600 non-null    str    
 2   churn_date                600 non-null    str    
 3   reason_code               600 non-null    str    
 4   refund_amount_usd         600 non-null    float64
 5   preceding_upgrade_flag    600 non-null    bool   
 6   preceding_downgrade_flag  600 non-null    bool   
 7   is_reactivation           600 non-null    bool   
 8   feedback_text             452 non-null    str    
dtypes: bool(3), float64(1), str(5)
memory usage: 30.0 KB


### Basic stats

In [32]:
accounts.describe(include="str")

,account_id,account_name,industry,country,signup_date,referral_source,plan_tier
count,500,500,500,500,500,500,500
unique,500,500,5,7,370,5,3
top,A-2e4581,Company_0,DevTools,US,2024-10-27,organic,Pro
freq,1,1,113,291,4,114,178


In [25]:
subscriptions.describe(include='str')

,subscription_id,account_id,start_date,end_date,plan_tier,billing_frequency
count,5000,5000,5000,486,5000,5000
unique,5000,500,625,208,3,2
top,S-8cec59,A-d4ac0e,2024-12-31,2024-12-31,Enterprise,monthly
freq,1,19,45,24,1723,2539


In [26]:
feature_usage.describe(include='str')

,usage_id,subscription_id,usage_date,feature_name
count,25000,25000,25000,25000
unique,24979,4967,731,40
top,U-25b56c,S-0896f4,2024-10-11,feature_12
freq,2,16,53,659


In [29]:
support_tickets.describe(include='str')

,ticket_id,account_id,submitted_at,closed_at,priority
count,2000,2000,2000,2000,2000
unique,2000,492,671,1893,4
top,T-0024de,A-bb3bd4,2024-12-01,2024-04-28 03:00:00,urgent
freq,1,11,10,3,514


In [30]:
churn_events.describe(include='str')

,churn_event_id,account_id,churn_date,reason_code,feedback_text
count,600,600,600,600,452
unique,600,352,311,6,3
top,C-816288,A-0a62f5,2024-12-29,features,too expensive
freq,1,5,7,114,161


### Null values

In [35]:
accounts.isna().sum()

account_id         0
account_name       0
industry           0
country            0
signup_date        0
referral_source    0
plan_tier          0
seats              0
is_trial           0
churn_flag         0
dtype: int64

In [36]:
subscriptions.isna().sum()

subscription_id         0
account_id              0
start_date              0
end_date             4514
plan_tier               0
seats                   0
mrr_amount              0
arr_amount              0
is_trial                0
upgrade_flag            0
downgrade_flag          0
churn_flag              0
billing_frequency       0
auto_renew_flag         0
dtype: int64

In [37]:
feature_usage.isna().sum()

usage_id               0
subscription_id        0
usage_date             0
feature_name           0
usage_count            0
usage_duration_secs    0
error_count            0
is_beta_feature        0
dtype: int64

In [38]:
support_tickets.isna().sum()

ticket_id                        0
account_id                       0
submitted_at                     0
closed_at                        0
resolution_time_hours            0
priority                         0
first_response_time_minutes      0
satisfaction_score             825
escalation_flag                  0
dtype: int64

In [39]:
churn_events.isna().sum()

churn_event_id                0
account_id                    0
churn_date                    0
reason_code                   0
refund_amount_usd             0
preceding_upgrade_flag        0
preceding_downgrade_flag      0
is_reactivation               0
feedback_text               148
dtype: int64

### Duplicate values

In [43]:
print(accounts.duplicated().sum())
print(subscriptions.duplicated().sum())
print(feature_usage.duplicated().sum())
print(support_tickets.duplicated().sum())
print(churn_events.duplicated().sum())

0
0
0
0
0


In [45]:
print(accounts["account_id"].duplicated().sum())
print(subscriptions["subscription_id"].duplicated().sum())
print(feature_usage["usage_id"].duplicated().sum())
print(support_tickets["ticket_id"].duplicated().sum())
print(churn_events["churn_event_id"].duplicated().sum())

0
0
21
0
0


In [66]:
print(subscriptions['account_id'].duplicated().sum())
print(feature_usage['subscription_id'].duplicated().sum())
print(support_tickets['account_id'].duplicated().sum())
print(churn_events['account_id'].duplicated().sum())

4500
20033
1508
248


### Cardinality

In [46]:
accounts['industry'].value_counts()

industry
DevTools         113
FinTech          112
Cybersecurity    100
HealthTech        96
EdTech            79
Name: count, dtype: int64

In [47]:
accounts['plan_tier'].value_counts()

plan_tier
Pro           178
Basic         168
Enterprise    154
Name: count, dtype: int64

In [48]:
accounts['referral_source'].value_counts()

referral_source
organic    114
other      103
ads         98
event       96
partner     89
Name: count, dtype: int64

In [49]:
accounts['country'].value_counts()

country
US    291
UK     58
IN     49
AU     32
DE     25
CA     23
FR     22
Name: count, dtype: int64

In [61]:
accounts['churn_flag'].value_counts()

churn_flag
False    390
True     110
Name: count, dtype: int64

In [51]:
subscriptions['plan_tier'].value_counts()

plan_tier
Enterprise    1723
Pro           1675
Basic         1602
Name: count, dtype: int64

In [60]:
subscriptions['churn_flag'].value_counts()

churn_flag
False    4514
True      486
Name: count, dtype: int64

In [52]:
feature_usage['feature_name'].value_counts()

feature_name
feature_12    659
feature_32    659
feature_6     655
feature_17    651
feature_34    650
feature_26    649
feature_36    648
feature_31    644
feature_20    643
feature_24    643
feature_38    643
feature_11    643
feature_2     642
feature_15    640
feature_29    640
feature_37    636
feature_22    636
feature_33    635
feature_10    634
feature_39    631
feature_30    629
feature_1     629
feature_4     625
feature_9     624
feature_28    623
feature_16    622
feature_40    621
feature_7     619
feature_25    615
feature_13    612
feature_27    611
feature_14    600
feature_3     598
feature_8     596
feature_21    591
feature_19    589
feature_18    588
feature_35    587
feature_5     577
feature_23    563
Name: count, dtype: int64

In [54]:
support_tickets['priority'].value_counts()

priority
urgent    514
high      510
medium    491
low       485
Name: count, dtype: int64

In [55]:
churn_events['reason_code'].value_counts()

reason_code
features      114
support       104
budget        104
unknown        95
competitor     92
pricing        91
Name: count, dtype: int64

### Relationships checks

Does every subscription belong to an account?

In [56]:
subscriptions["account_id"].isin(accounts["account_id"]).all()

np.True_

Does every feature usage belong to a subscription?

In [57]:
feature_usage["subscription_id"].isin(subscriptions["subscription_id"]).all()

np.True_

Does every support ticket belong to an account?

In [58]:
support_tickets["account_id"].isin(accounts["account_id"]).all()

np.True_

Does every churn event belong to an account?

In [59]:
churn_events["account_id"].isin(accounts["account_id"]).all()

np.True_

# Data Understanding Findings

## General Observations

1. The shape of every table matches the dataset documentation, indicating that no records are missing during the import process.

2. Trial subscriptions have both **MRR (Monthly Recurring Revenue)** and **ARR (Annual Recurring Revenue)** equal to **0**, which is consistent with the business logic since trial users do not generate revenue.

3. All date columns were imported as strings (`object` type in Pandas) and will need to be converted to proper datetime objects before performing any temporal analysis.

## Subscription Table

5. Out of **5,000** subscriptions, **4,514** are currently active because their `end_date` is null. Only **486** subscriptions have ended.

6. An account can appear multiple times in the `subscriptions` table, meaning a customer may have multiple subscription periods due to upgrades, downgrades, renewals, or plan changes.

## Support Tickets

7. The `satisfaction_score` column contains **825 missing values** out of **2,000** support tickets. These missing values likely represent customers who did not respond to the satisfaction survey rather than missing data caused by collection errors.

## Churn Events

8. The `feedback_text` column contains **148 customer comments** out of **600** churn events. This text could be used in a future extension of the project using Natural Language Processing (NLP) techniques such as sentiment analysis or keyword extraction.

9. There are **six** churn reason categories:

   * Features
   * Support
   * Budget
   * Unknown
   * Competitor
   * Pricing

## Feature Usage

10. The dataset documentation states that RavenStack provides **40 different product features**, which matches the observed data.

11. The `feature_usage` table contains **21 duplicated values** in the `usage_id` column, even though `usage_id` is documented as the primary key. This anomaly requires further investigation before the data is used for analysis.

## Accounts

12. The `accounts` table contains **500** customers, of which **110** have churned at some point, corresponding to a churn rate of approximately **22%**.

13. The dataset includes:

* 5 industry categories
* 3 subscription plans
* 5 referral sources
* 7 countries

## Relationships

14. The relationships between tables were successfully validated:

* Every subscription belongs to an existing account.
* Every feature usage record belongs to an existing subscription.
* Every support ticket belongs to an existing account.
* Every churn event belongs to an existing account.

15. The data follows the expected business granularity:

* One account can have multiple subscriptions.
* One subscription can generate multiple feature usage events.
* One account can generate multiple support tickets.
* One account can experience one or more churn-related events.


### Usage ID duplicated problem

In [67]:
duplicates = feature_usage[
    feature_usage["usage_id"].duplicated(keep=False)
].sort_values("usage_id")

duplicates

,usage_id,subscription_id,usage_date,feature_name,usage_count,usage_duration_secs,error_count,is_beta_feature
19294,U-0c9318,S-01b2dc,2023-12-11,feature_9,5,1060,3,False
17533,U-0c9318,S-0ffab0,2024-01-30,feature_11,3,1614,0,False
20588,U-13ce5b,S-9b623b,2023-03-26,feature_28,10,2050,1,False
9626,U-13ce5b,S-8b0950,2024-09-10,feature_9,8,824,0,True
18480,U-2103bb,S-7fc49b,2023-04-18,feature_12,9,5022,1,False
10379,U-2103bb,S-ae3270,2024-01-06,feature_3,10,5510,0,False
22,U-25b56c,S-810c27,2024-10-06,feature_20,7,231,0,False
7574,U-25b56c,S-34253c,2023-10-28,feature_20,6,2166,1,False
21376,U-48a4aa,S-93f835,2024-08-24,feature_39,10,4770,0,False
1085,U-48a4aa,S-383ac2,2023-02-12,feature_28,14,7126,0,False


In [68]:
duplicates["usage_id"].value_counts()

usage_id
U-0c9318    2
U-13ce5b    2
U-2103bb    2
U-25b56c    2
U-48a4aa    2
U-4ef9da    2
U-52f4ae    2
U-5cd3c2    2
U-649cec    2
U-662254    2
U-6970d1    2
U-74b23f    2
U-7a87ed    2
U-7c6703    2
U-8128da    2
U-8a3738    2
U-ae1a6c    2
U-b0ae95    2
U-c878da    2
U-d20c6d    2
U-f42b0a    2
Name: count, dtype: int64

In [71]:
duplicate_rows = feature_usage["usage_id"].duplicated().sum()
total_rows = len(feature_usage)

print(f"{duplicate_rows} duplicated IDs out of {total_rows} rows")
print(f"{duplicate_rows / total_rows:.4%} of the dataset")

21 duplicated IDs out of 25000 rows
0.0840% of the dataset


The feature_usage table contains 21 duplicated primary key values. Since these rows represent distinct usage events rather than duplicated observations, the duplicated identifiers appear to be a data generation issue. In a specific module, new unique identifiers will be generated while preserving the original raw dataset.